server/src/main/java/io/crate/expression/scalar/HasTablePrivilegeFunction.java

```diff
@@ -0,0 +1,178 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import static io.crate.metadata.Scalar.DETERMINISTIC_ONLY;
+
+import java.util.Collection;
+
+import io.crate.metadata.FunctionName;
+import io.crate.metadata.Functions;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.Schemas;
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogSchemaInfo;
+import io.crate.role.Permission;
+import io.crate.role.Role;
+import io.crate.role.Roles;
+import io.crate.role.Securable;
+import io.crate.types.DataTypes;
+
+public class HasTablePrivilegeFunction {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "has_table_privilege");
+
+    public static boolean checkByTableName(Roles roles, Role user, Object table, Collection<Permission> permissions, Schemas schemas) {
+        String tableFqn = RelationName.fqnFromIndexName((String) table);
+        return checkByTablePrivileges(roles, user, tableFqn, permissions);
+    }
+
+    public static boolean checkByTableOid(Roles roles, Role user, Object table, Collection<Permission> permissions, Schemas schemas) {
+        int tableOid = (int) table;
+        RelationName relationName = schemas.getRelation(tableOid);
+        if (relationName == null) {
+            throw new IllegalArgumentException("Cannot find corresponding relation by the given oid");
+        }
+        String tableFqn = relationName.fqn();
+        return checkByTablePrivileges(roles, user, tableFqn, permissions);
+    }
+
+    private static boolean checkByTablePrivileges(Roles roles, Role user, String table, Collection<Permission> permissions) {
+        boolean result = false;
+        for (Permission permission : permissions) {
+            result |= roles.hasPrivilege(user, permission, Securable.TABLE, table);
+        }
+        if (!result) {
+            for (Permission permission : permissions) {
+                result |= roles.hasPrivilege(user, permission, Securable.VIEW, table);
+            }
+        }
+        return result;
+    }
+
+    public static void register(Functions.Builder module) {
+        // Signature without user, takes user from session.
+        module.add(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(), // Table
+                DataTypes.STRING.getTypeSignature(), // Privilege
+                DataTypes.BOOLEAN.getTypeSignature()
+            ).withFeatures(DETERMINISTIC_ONLY),
+            (signature, boundSignature) -> new HasPrivilegeFunction(
+                signature,
+                boundSignature,
+                HasPrivilegeFunction::userByName,
+                HasTablePrivilegeFunction::checkByTableName,
+                HasSchemaPrivilegeFunction::parsePermissions
```

Comment:
The has_table_privilege needs special `parsePermissions`. From Postgres docs:
```
has_table_privilege ( [ user name or oid, ] table text or oid, privilege text ) → boolean

Does user have privilege for table? Allowable privilege types are SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, and TRIGGER.
```
https://www.postgresql.org/docs/16/functions-info.html#FUNCTIONS-INFO-ACCESS

Not sure how to treat the last three, maybe return false always for those 3 types? (unless of course table doesn't exist).

Reply:
Tests need to be modified to test those privilege types and not the ones for `has_schema_privilege`

Reply:
> maybe return false always for those 3 types?

How about throwing? Since we throw here as well:
```
cr> grant truncate on table q to john;                                                                                                                                
SQLParseException[Unknown permission 'truncate']
```

Reply:
Also good, yes.